### PR TITLE
One resource entry per resource

### PR DESCRIPTION
**Pull request by plehegar**
This comes from https://github.com/w3c/web-platform-tests/pull/2884 and this test scenario needs to be addressed by the specification first.

### CODE REVIEW

**Comment by plehegar** [comment]
@igrigorik do you know where we stand on this front? Is the test appropriate and is the case covered by the spec?

**Comment by igrigorik** [comment]
```
 +    observer.observe({entryTypes: ["resource"]});
 +    var img = document.createElement("img");
 +    img.src = img_location;
 +    document.body.appendChild(img);
 +    img = document.createElement("img");
 +    img.src = img_location;
 +    document.body.appendChild(img);
 +  }, "Only one resource entry per resource");
```

It's still not clear to me if this is correct. If you follow the RT processing model verbatim, then there should be multiple records, since the second image should (at least in theory) initiate a fetch, which should then hit the appropriate caches and pull out the right data. 

However, in practice, in Chrome you only see one request because the request is fulfilled via the ~"renderer memory cache", which is not specified anywhere... I think we should solve that first (probably within Fetch), and then we'll be in a position to answer this.

/cc @toddreifsteck @annevk

**Comment by toddreifsteck** [comment]
Said another way.. if there was a ServiceWorker behind the fetches, you'd see 2 entries. If there isn't, you may only see 1 today. This is probably not correct and we should consider updating the spec. (We will also need to keep in mind that web devs are used to this behavior today and changing this could break some telemetry processing.)

**Comment by annevk** [comment]
HTML defines an image cache. Forgot the name atm but I think that is defined to some extent. Not well, since not subject to CSP atm, but…

**Comment by gsnedders** [comment]
```
15:01 < aryehgregor> Um, I might have done something wrong by mistake.
15:01 < aryehgregor> I think I pushed to the wrong repo?
```

File: resource-timing/po-resource-single.html [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+<script>
+  async_test(function (t) {
+    function path(pathname) {
+      var filename = pathname.substring(pathname.lastIndexOf('/')+1);
+      return pathname.substring(0, pathname.length - filename.length);
+    }
+    var img_location = document.location.origin + path(document.location.pathname)
+       + "resources/square.png";
+    var observer = new PerformanceObserver(
+        t.step_func(function (entryList, obs) {
```
Comment (yoavweiss) [inline resource-timing/po-resource-single.html]: We are not testing here a scenario where there will be a second entry added, only in some slight delay that will cause it to be added to the next PerfObs event.

**Comment by yoavweiss** [comment]
> HTML defines an image cache. Forgot the name atm but I think that is defined to some extent. Not well, since not subject to CSP atm, but…

@annevk - I think you're referring to [the list of available images](https://html.spec.whatwg.org/multipage/embedded-content.html#the-list-of-available-images).

In any case, I disagree that this case should show two separate entries, as there's no actual fetch triggered for MemoryCache reused images. I'd argue that if we want 2 separate entries here, we should also have 2 entries for every resource the preloader is triggering a preload of. There's no conceptual difference between the two.

> Said another way.. if there was a ServiceWorker behind the fetches, you'd see 2 entries. If there isn't, you may only see 1 today. This is probably not correct and we should consider updating the spec.

@toddreifsteck - Why is that? Can you point to the relevant parts of the SW/fetch specs that result in resource reuse triggering new fetch events?

**Comment by annevk** [comment]
> In any case, I disagree that this case should show two separate entries, as there's no actual fetch triggered for MemoryCache reused images. I'd argue that if we want 2 separate entries here, we should also have 2 entries for every resource the preloader is triggering a preload of. There's no conceptual difference between the two.

When are you planning on getting this defined in detail?

**Comment by igrigorik** [comment]
Discussed this at TPAC yesterday with Todd & Yoav...

> It's still not clear to me if this is correct. If you follow the RT processing model verbatim, then there should be multiple records, since the second image should (at least in theory) initiate a fetch, which should then hit the appropriate caches and pull out the right data.

Ignore the above. [Step 6](https://w3c.github.io/resource-timing/#processing-model) of the RT processing model covers this case:

> If the user agent is to reuse the data from another existing or completed fetch initiated from the current document, abort the remaining steps.

So, the behavior that is being tested here *is correct* -- lgtm for that. For the test structure: it's not clear to me that this is the best way to test the actual behavior, as there is no guarantee that UA must dispatch a single PO event. Ideally, we should check that only one event was delivered some time after the onload event of the second test fired.

**Comment by toddreifsteck** [comment]
@plehegar Is updating this test per @igrigorik feedback something you can take or should we move it to Xiaoqian?

**Comment by igrigorik** [comment]
@siusin is this something you'd be able to update (see https://github.com/w3c/web-platform-tests/pull/2919#issuecomment-248717176) while you're working on https://github.com/w3c/web-platform-tests/pull/402?

**Comment by plehegar** [comment]
Sorry, I missed that earlier :(
To be clear, the proposal is to change the test as follows:
1- attach an observer before the load event
2- add two img tags in the markup, pointing to the same location
3- once the load event fired, check that we only got one entry for the image?

**Comment by igrigorik** [comment]
Yep! For (1), make sure we register it at the top of the page, well ahead of the images..

**Comment by plehegar** [comment]
not sure why Chrome is showing timeouts here. I get pass on chrome dev and canary. I'll keep thinking about this test and whether it should get adjusted...

**Comment by igrigorik** [comment]
Do we have a URL to debug? http://w3c-test.org/submissions/2919/resource-timing/po-resource-single.html doesn't seem to work.

**Comment by jgraham** [comment]
Per the travis logs it's pretty clear that the Firefox result above really is Firefox and the results aren't confused.

**Comment by plehegar** [comment]
w3c-test:mirror

**Comment by plehegar** [comment]
@igrigorik done. @jgraham thanks

**Comment by plehegar** [comment]
ok, this is a mystery why the test keeps getting a timeout in the CI and I don't get that locally :(

**Comment by plehegar** [comment]
(the pull request needs to be rebased)

**Comment by plehegar** [comment]
At this point, I'm thinking to wait for https://github.com/w3c/resource-timing/pull/83 ...

**Comment by igrigorik** [comment]
@plehegar wait for...? As in, tests to cover https://github.com/w3c/resource-timing/issues/87?

**Comment by plehegar** [comment]
No. I did mean https://github.com/w3c/resource-timing/pull/83 . For the purpose of this test, I need to be able to better predict when the entry will be reported and that's not the case with Firefox at the moment.

**Comment by igrigorik** [comment]
Err, still confused. What does responseStart have to do with when entry is reported (I assume you mean queued to PerfObserver)?

**Comment by igrigorik** [comment]
@siusin any chance we can stage this test?

**Comment by yoavweiss** [comment]
Resubmitted at https://github.com/w3c/web-platform-tests/pull/6567

**Comment by igrigorik** [comment]
Thanks Yoav. Closing this, let's continue in the new PR.

**Pull request closed without merging**